In [84]:

#import Required Libraries

In [85]:
import numpy as np
import pandas as pd
import itertools
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.linear_model import PassiveAggressiveClassifier
from sklearn.metrics import accuracy_score, confusion_matrix

In [86]:
#Retrieve the data

In [87]:
#what is dataframe ?

df = pd.read_csv('news.csv')

#get shape and head
df.shape             # dimesion of dataset
df.head(n=5)         # return the first n rows of dataframe, default n=5. tail is the opposite

Unnamed: 0                                              title  \
0        8476                       You Can Smell Hillary’s Fear   
1       10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2        3608        Kerry to go to Paris in gesture of sympathy   
3       10142  Bernie supporters on Twitter erupt in anger ag...   
4         875   The Battle of New York: Why This Primary Matters   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4  It's primary day in New York and front-runners...  REAL

In [88]:
#getting labels
labels = df.label         #get the column named 'label'
labels.head()             #print label first 5 rows

0    FAKE
1    FAKE
2    REAL
3    FAKE
4    REAL
Name: label, dtype: object

In [89]:
#creating training, testing sets
x_train,x_test,y_train,y_test=train_test_split(df['text'], labels, test_size=0.2, random_state=7)           #split dataset into training and testing
print( x_train.shape)           #print training set
print( x_train.shape)           #print testing set


(5068,)
(5068,)


In [90]:
# what is stop words ?
# what is TfidfVectorizer ?
# what is sparsed array ?

# Initialize TfidfVectorizer
tfidf_vectorizer=TfidfVectorizer(stop_words='english', max_df=0.7)


tfidf_train = tfidf_vectorizer.fit_transform(x_train)         # fit and transform training set
tfidf_test = tfidf_vectorizer.transform(x_test)               # transform test set
#print(x_train.head())
#print('Vectorized')
#print(tfidf_train)



In [91]:
# What is PassiveAggressiveClassifier


pac = PassiveAggressiveClassifier(max_iter=50)          #initalize PassiveAggressiveClassifier
pac.fit(tfidf_train, y_train)


y_pred = pac.predict(tfidf_test)                        #Predict on test set
score=accuracy_score(y_test,y_pred)                     #Calculate Accuracy
print(f'Accuracy: {round(score*100,2)}%')

Accuracy: 92.82%


In [92]:
# What is Confusion_matrix ?

confusion_matrix(y_test, y_pred, labels=['FAKE','REAL'])


array([[590,  48],
       [ 43, 586]], dtype=int64)

<dl>
     <dt>So with this model</dt>
     <dd>we have 589 true positives,</dd> 
     <dd>587 true negatives,</dd>
     <dd>42 false positives, and</dd>
     <dd>49 false negatives.</dd>
</dl>


## Summary

Today, we learned to detect fake news with Python. We took a political dataset, implemented a TfidfVectorizer, initialized a PassiveAggressiveClassifier, and fit our model. We ended up obtaining an accuracy of 92.82% in magnitude.

Hope you enjoyed our first advanced level Python Project. Keep visiting DataFlair for more interesting Python data science and machine learning projects.